### PR TITLE
MM-54173: Part 2 - unify at mention components

### PR DESCRIPTION
**Pull request by StreakInTheSky**
#### Summary

A continuation to PR #24402. Like the tickets mentions, wanted "to reuse whatever function normal at mentions is using to create the linking to handle group mentions." So I merged the two components for user at mentions and group mentions. A bit trickier than originally anticipated. Here are the major changes.

- Changed the original class component to the functional component used for the Groups component.
- Factored out a couple utility functions and added tests for them.
   -  `approxGroupPopOverHeight` gets the approximate height of the groups popover.
   -  `getUserOrGroupFromMentionName` figures out whether the mention should be for a user and a group . There was originally another function with that same name that I renamed to `getMentionDetails`, which I thought made more sense. Added tests to both.

Most of the logic for the functions were there, just took them out as standalone functions.

Also because of the change to a functional components I ended up removing some Enzyme tests. They tested for the `placement` state of the popover. Tried to re-implement the tests, but that proved to be too difficult. I came to the conclusion that they weren't actually necessary because they were just testing the component's state. That state was actually coming from a utility function, `popOverOverlayPosition`, that was already being tested for the correct output.

Also misc small changes to clean up the code and handle discriminating unions.

#### Release Note

```release-note
NONE
```

### CODE REVIEW

**Comment by larkox** [comment]
We have merged https://github.com/mattermost/mattermost/pull/24091, which adds the "import order" lint rule. This PR touches almost every file under `webapp/channels/src` folder, so you will most surely get a bunch of conflicts. So... a few comments here:
- This PR should only introduce conflicts on the imports area. Anything outside of the imports is something you have to look more closely, since it could be from other PRs.
- My recommendation for solving the conflicts is: accept both changes and delete the duplicates. This allow you to verify both are importing from the same place and that stuff that was moved is still present.
- You will probably need to run `npm run fix` to reorder back the imports after the conflict resolution. Not often, but sometimes you may need to run it twice.

If you have any trouble with the merge, do not hesitate to ask me.

File: webapp/channels/src/components/at_mention/at_mention.tsx
Comment (enahum) [inline webapp/channels/src/components/at_mention/at_mention.tsx:113]: why an empty string? does it makes sense to have it as `undefined` instead?

File: webapp/channels/src/components/at_mention/at_mention.tsx
Comment (enahum) [inline webapp/channels/src/components/at_mention/at_mention.tsx:138]: should this be returning null instead ? an empty string is not a component as the two previous returns are

File: webapp/channels/src/components/at_mention/at_mention.tsx
Comment (enahum) [inline webapp/channels/src/components/at_mention/at_mention.tsx:191]:
```suggestion
                    ) : null
```

File: webapp/channels/src/utils/post_utils.ts
Comment (enahum) [inline webapp/channels/src/utils/post_utils.ts:732]: I rather be `null` or `undefined` instead of an empty string

**Review by enahum — CHANGES_REQUESTED** [review]

File: webapp/channels/src/utils/post_utils.ts
Comment (StreakInTheSky) [reply webapp/channels/src/utils/post_utils.ts:732]: Ok. I used an empty string since that's what the original logic was using. Didn't know if it was some convention or something. Would have preferred to use `null`, myself.

File: webapp/channels/src/components/at_mention/at_mention.tsx
Comment (StreakInTheSky) [reply webapp/channels/src/components/at_mention/at_mention.tsx:74]: So I'm not actually super familiar with using `useCallback`, but the handlers are used on an `a` tag, not on a child component, would the hook still be useful in that case?

File: webapp/channels/src/utils/post_utils.ts
Comment (StreakInTheSky) [reply webapp/channels/src/utils/post_utils.ts]: ended up using `undefined` so didn't have to use explicit unions everywhere

**Comment by StreakInTheSky** [comment]
bump

File: webapp/channels/src/components/at_mention/at_mention.test.tsx
Comment (enahum) [inline webapp/channels/src/components/at_mention/at_mention.test.tsx:231]: Why the removal of the placement tests? We should be testing for this to make sure it still opens correctly

File: webapp/channels/src/components/at_mention/at_mention.tsx
Comment (enahum) [inline webapp/channels/src/components/at_mention/at_mention.tsx:141]: You should probably memoize this

**Review by enahum — CHANGES_REQUESTED** [review]

**Comment by enahum** [comment]
You also need to fix the check failures

File: webapp/channels/src/components/at_mention/at_mention.test.tsx
Comment (StreakInTheSky) [reply webapp/channels/src/components/at_mention/at_mention.test.tsx:231]: The test doesn't work with functional components with React hooks. I tried to see if I could update it for hooks, but was spending way too much time on it without any result. It was really just testing whether the component's `placement` state was updated to the correct string. That string comes from a utility function, which wasn't tested. So instead of banging my head on testing for updated state, created tests for making sure the utility function outputs the correct string.

**Comment by enahum** [comment]
tests are failing

**Comment by StreakInTheSky** [comment]
> tests are failing

ah looks like some snapshot tests with features from master. Merging master and will look into it.

**Comment by StreakInTheSky** [comment]
Bump

**Review by enahum — APPROVED** [review]

**Comment by StreakInTheSky** [comment]
@calebroseland Hi, looks like this is still waiting for your approval? If there's anything you think still need to be done/fixed, let me know. Thanks!

**Review by calebroseland — CHANGES_REQUESTED** [review]
Hi @StreakInTheSky, please accept my apologies for my delay in reviewing this.

Overall looks pretty good for combining the two components, just the one request for the the `null` child in `<Overlay>` issue.


(Additionally, thoughts for later)
A few followup changes in `at_mention` (re: preexisting logic from `at_mention_group`) might be good:
1. `target` as a separate piece of state isn't strictly needed if the ref is already there, then `target` wouldn't have to be set or passed around in the functions
2. `const [user, group] = //...` could go at the top of the component—then `group` wouldn't have to be passed around in the functions
3. the dual `Overlay`s seems unnecessary to me, it might be nicer to simplify to just one Overlay

Perhaps something like:
```tsx
//...
{ref.current && (
    <Overlay
        placement={placement}
        show={show}
        target={ref.current}
        rootClose={true}
        onHide={hideOverlay}
        onExited={onHidden}
        shouldUpdatePosition={true}
    >
        {getOverlayContent(user, group, groupUser)}
    </Overlay>
)}
<a
//...
```

File: webapp/channels/src/components/at_mention/at_mention.tsx
Comment (StreakInTheSky) [reply webapp/channels/src/components/at_mention/at_mention.tsx:194]: oh interesting, wasn't aware of this issue. Noticed the `span` before and thought it wasn't necessary. Will put it back.

**Comment by StreakInTheSky** [comment]
So I put the span back and documented why it's there. Also moved the `const [user, group] = ...` part near the top of the component and cleaned up passing `group` through the functions.

Looked into the other things you mentioned and looks like they're all kind of related. Changed it to use only one `Overlay`, but don't have a way to test if it's actually working as intended since it's an enterprise feature and can't really write unit tests to test it. I'll make a new PR for it, so this can be merged?

**Review by calebroseland — APPROVED** [review]
> I'll make a new PR for it, so this can be merged?

Yep, sounds good 👍

**Comment by lindy65** [comment]
Hi @calebroseland - is there a Jira ticket for this PR?  I couldn't find one and the PR referred to at the top of this PR is merged. Not sure whether testing is the same as on the closed Jira ticket?

**Comment by calebroseland** [comment]
@lindy65, I don't think there is one, but QA steps based on the code changes for this would be:

Ensure at-mention profile, group, and group-member popovers work in posts:
1. User at-mention profile popover appears
CleanShot 2023-10-12 at 15 39 20@2x

2. Group at-mention popover appears
CleanShot 2023-10-12 at 15 39 25@2x


3. Group at-mention -> group-member popover appears (e.g. after clicking on a group member from group at-mention popover)
CleanShot 2023-10-12 at 15 39 32@2x

**Comment by lindy65** [comment]
Thanks @calebroseland - I've created https://mattermost.atlassian.net/browse/MM-54932 to track post merge

**Review by lindy65 — APPROVED** [review]
Thanks @StreakInTheSky - looks good to me 👍 

cc @calebroseland

**Comment by StreakInTheSky** [comment]
Hi, wondering if this can be merged? Thanks!

**Comment by lindy65** [comment]
> Hi, wondering if this can be merged? Thanks!

cc @enahum @calebroseland

**Pull request merged**
